Use bitvector types to implement __builtin_{add,sub,mul}_overflow

tautschnig opened the pull request:
The use of mathematical integers is not supported by the SAT back-end.
Although GCC's specification describes the operations over mathematical
integers, the same precision can be achieved by using bitvector types of
sufficient width.

The implementation now also introduces a fresh symbol to store the
result of the full-precision operation. This ensures that the arguments
to __builtin_{add,sub,mul}_overflow are evaluated exactly once.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Implement GCC's __builtin_{add,sub,mul}_overflow_p

Refactor and expand the implementation of
__builtin_{add,sub,mul}_overflow (i.e., without the _p suffix) to also
support the variant that does not store the computed result.

kroening commented:
Note that we've got expressions ID_overflow_plus, ID_overflow_minus and ID_overflow_mult -- these make attempts at optimizing the encoding required for these predicates.

tautschnig commented:
> Note that we've got expressions ID_overflow_plus, ID_overflow_minus and ID_overflow_mult -- these make attempts at optimizing the encoding required for these predicates.

I am well aware of those, and did wonder whether we can just use them instead of making a big fuzz here. My concern was twofold, and I'd appreciate further thoughts on this:

1) What type should be used for the operands? `__builtin_{add,sub,mul}_overflow` permit using three different types.
2) `__builtin_{add,sub,mul}_overflow` require computing/storing the result, not just performing the overflow check. Thus, preferably, the actual arithmetic operation should only be done once.

kroening commented:
1. `ID_overflow_{plus, minus, mult}` support the three types needed for `__builtin_{add,sub,mul}_overflow`.

2. There is a bit of savings potential when doing the overflow check for multiplication given the way it's implemented now. No significant savings when doing plus/minus. There is the option to do something iterative:

https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/z3prefix.pdf

tautschnig commented:
> `ID_overflow_{plus, minus, mult}` support the three types needed for `__builtin_{add,sub,mul}_overflow`.

Where would that be supported? `boolbvt::convert_overflow` appears to assume that the two operands have the same type.

kroening commented:
https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html appears to suggest that the two operands do indeed have the same type.

tautschnig commented:
> https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html appears to suggest that the two operands do indeed have the same type.

I'm not sure I can see this? The signature `bool __builtin_add_overflow (type1 a, type2 b, type3 *res)` doesn't seem to require this? The comment also says "The first built-in function allows arbitrary integral types for operands and the result type must be pointer to some integral type other than enumerated or boolean type, [...]"

tautschnig committed:
Use bitvector types to implement __builtin_{add,sub,mul}_overflow

The use of mathematical integers is not supported by the SAT back-end.
Although GCC's specification describes the operations over mathematical
integers, the same precision can be achieved by using bitvector types of
sufficient width.

The implementation now also introduces a fresh symbol to store the
result of the full-precision operation. This ensures that the arguments
to __builtin_{add,sub,mul}_overflow are evaluated exactly once.
